include/lo2s/perf/event.hpp

```diff
@@ -0,0 +1,288 @@
+    {
+        attr_.exclude_kernel = exclude_kernel;
+    }
+
+    void set_sample_period(const int& period);
+    void set_sample_freq(const uint64_t& freq);
+    void event_attr_update(std::uint64_t value, const std::string& format);
+
+    void parse_pmu_path(const std::string& ev_name);
+    void parse_cpus();
+    const std::set<Cpu>& supported_cpus() const;
+
+    bool is_valid() const;
+    bool event_is_openable();
+
+    bool is_available_in(ExecutionScope scope) const
+    {
+        // per-process should always work. the counter will just not count if the process is
+        // scheduled on a core that is not supprted by that counter
+        return scope.is_thread() || cpus_.empty() || cpus_.count(scope.as_cpu());
+    }
+
+    bool degrade_precision();
+
+    friend bool operator==(const Event& lhs, const Event& rhs)
+    {
+        return !memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+    friend bool operator<(const Event& lhs, const Event& rhs)
+    {
+        return memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+    friend bool operator>(const Event& lhs, const Event& rhs)
+    {
+        return memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+protected:
+    void update_availability();
+    void set_common_attrs(bool enable_on_exec);
+
+    struct perf_event_attr attr_;
+    double scale_ = 1;
+    std::string unit_ = "#";
+
+    std::string name_ = "";
+    std::set<Cpu> cpus_;
+    Availability availability_ = Availability::UNAVAILABLE;
+
+    std::filesystem::path pmu_path_;
+    std::string pmu_name_;
+};
+
+/**
+ * Contains an event parsed from sysfs
+ * @note call on use_sampling_options() after creation to get a valid
+ * event, otherwise the availability will be set to UNAVAILABLE
+ */
+class SysfsEvent : public Event
+{
+public:
+    using Event::Event;
+    SysfsEvent(const std::string& ev_name, bool enable_on_exec = false);
+
+    void use_sampling_options(const bool& use_pebs, const bool& sampling, const bool& enable_cct);
+};
+
+/**
+ * Contains an opened instance of Event.
+ * Use any Event.open() method to construct an object
+ */
+class EventGuard
+{
+public:
+    EventGuard();
+    EventGuard(Event& ev, std::variant<Cpu, Thread> location, int group_fd, int cgroup_fd);
+
+    EventGuard(EventGuard&) = delete;
+    EventGuard& operator=(const EventGuard&) = delete;
+
+    EventGuard(EventGuard&& other)
+    {
+        std::swap(fd_, other.fd_);
+        std::swap(ev_, other.ev_);
+    }
+
+    EventGuard& operator=(EventGuard&& other)
+    {
+        std::swap(fd_, other.fd_);
+        std::swap(ev_, other.ev_);
+        return *this;
+    }
+
+    /**
+     * opens child as a counter of the calling (leader) event
+     */
+    EventGuard open_child(Event child, ExecutionScope location, int cgroup_fd = -1);
+
+    void enable();
+    void disable();
+
+    void set_id(const uint64_t& id)
```

Comment:
This ioctl does not set the id, it  gets the id. From the `perf_event_open` manpage:

> This returns the event ID value for the given event file descriptor.  

please rewrite as uint64_t get_id() function